sklearn/mixture/tests/test_gaussian_mixture.py

```diff
@@ -785,6 +784,51 @@ def test_warm_start():
     assert_true(h.converged_)
 
 
+@ignore_warnings(category=ConvergenceWarning)
+def test_convergence_detected_with_warm_start():
+    # We check that convergence is detected when warm_start=True
+    rng = np.random.RandomState(0)
+    rand_data = RandomData(rng)
+    n_components = rand_data.n_components
+    X = rand_data.X['full']
+
+    for max_iter in (1, 2, 50):
+        gmm = GaussianMixture(n_components=n_components, warm_start=True,
+                              max_iter=max_iter, random_state=rng)
+        for _ in range(100):
+            gmm.fit(X)
+            if gmm.converged_:
+                break
+        assert gmm.converged_
+        assert max_iter >= gmm.n_iter_
+
+
+@ignore_warnings(category=ConvergenceWarning)
+def test_no_convergence_at_1st_iter_if_warm_start_and_max_iter_greater_1():
+    # When warm_start=True and max_iter > 1, if we change dataset, we don't
+    # want to detect convergence by mistake at the first iteration if
+    # lower_bound_ is very close to the lower_bound_ after the previous call
+    # to the fit method.
+    # Unlikely, but possible and problematic, so we might as well avoid it.
+    rng = np.random.RandomState(0)
```

Comment:
Sorry I'm being slow. I don't see where this case is handled and I don't understand the test.  Why do we always reset with max_iter > 1 now?

Reply:
No worries, let me explain. The scenario I'm trying to avoid, which could happen today (without this PR), is using `warm_start=True`, and you run `gm.fit(X1)` followed by `gm.fit(X2)`, where `X1` and `X2` are different datasets, and if you are unlucky the first iteration of `gm.fit(X2)` happens to compute a lower bound very close to the final lower bound of `gm.fit(X1)`, so the algorithm thinks the algorithm has converged when it fact it should have continued to iterate.  Sure, this is quite unlikely with the default `tol` and if `X1` and `X2` are very different, but it might be dangerously likely if `X1` and `X2` are very similar, but not identical (or if `tol` is high).
Since it is hard to find two datasets `X1` and `X2` where this scenario occurs, I test this scenario by setting `tol` to infinity.
This case is handled on line 217 of `base.py`: we start the iterations with `lower_bound = (-np.infty if do_init or self.max_iter > 1 else self.lower_bound_)`. So we only continue from the final lower bound of the last call to `fit()` if `warm_start` is `True` and `max_iter == 1`. The assumption is that people who use `warm_start=True` and `max_iter = 1` are certainly doing this to manually run the training loop themselves on a single dataset, but if they are using `max_iter > 1`, it is unclear whether they are running consecutive calls to `fit()` on the same dataset or not, so we should err on the safe side.
So in short there are two issues in the current implementation: (1) if `n_init > 1`, the lower bound is the one from the last initialization, not from the best initialization, and (2) there is a risk of false convergence if `warm_start` is `True`.  Since they are both due to the same few lines of code, I fixed them both in this one PR.
Hope this helps.